### ISSUE TITLE
POP and PUSH relocation types

### ISSUE THREAD

**Issue by maoif** [loongson/binutils-gdb#134]
I found that the relocation types of LoongArch contains pops and pushes, e.g., `R_LARCH_SOP_PUSH_PCREL`, and in BFD, `elfnn-loongarch.c`, there is a stack `larch_opc_stack`. However other architectures do not employ a stack to perform relocations.

Can any one explain the design of these relocation types? Does the order of relocation entries matter in loongarch binaries?

**Comment by xen0n** [comment]
Hi,

There *are* other architectures employing similar design for relocs, e.g. [`rx`](https://github.com/bminor/binutils-gdb/blob/binutils-2_38/bfd/elf32-rx.c#L438-L460) (I think LoongArch's implementation is just copied from there) and [`rl78`](https://github.com/bminor/binutils-gdb/blob/binutils-2_38/bfd/elf32-rl78.c#L362-L383). They are both niche architectures, and are from a single Japanese company, so the practice is definitely not known to wider audience, but not to say it's unfeasible.

Basically, one process the reloc records attached to the same address sequentially, essentially running a stack machine, to carry out the actual reloc operations. (Yes, the order totally matters.)

This design is flexible, in that only assemblers need to be aware of new reloc types, but the significant downside is (a) limited chances for parallelization and (b) huge bloat in object code. I have drafted a more traditional scheme of relocations, but I personally don't have enough time to push it across the finish line (I'm not Loongson employee, and not paid for my Loongson work at all) so I pushed Loongson to do the same. I hear they're working on their version of such a scheme, and they may have something to share in the coming days.

**Comment by xen0n** [comment]
For the implementation details: the macro instructions are expanded into open-coded snippets, defined in the assembler's opcodes array:

https://github.com/bminor/binutils-gdb/blob/binutils-2_38/opcodes/loongarch-opc.c#L147-L260

Then the expressions just get parsed, and emitted as stack operations that become the reloc records.

- [lexer](https://github.com/bminor/binutils-gdb/blob/binutils-2_38/opcodes/loongarch-opc.c#L147-L273)
- [parser](https://github.com/bminor/binutils-gdb/blob/binutils-2_38/gas/config/loongarch-parse.y)

Finally `loongarch_parse_expr` gets invoked by the target-specific code:

https://github.com/bminor/binutils-gdb/blob/binutils-2_38/gas/config/tc-loongarch.c#L593-L659

Seems not worth it, huh?

**Comment by maoif** [comment]
> For the implementation details: the macro instructions are expanded into open-coded snippets, defined in the assembler's opcodes array:
> 
> https://github.com/bminor/binutils-gdb/blob/binutils-2_38/opcodes/loongarch-opc.c#L147-L260
> 
> Then the expressions just get parsed, and emitted as stack operations that become the reloc records.
> 
>     * [lexer](https://github.com/bminor/binutils-gdb/blob/binutils-2_38/opcodes/loongarch-opc.c#L147-L273)
> 
>     * [parser](https://github.com/bminor/binutils-gdb/blob/binutils-2_38/gas/config/loongarch-parse.y)
> 
> 
> This is all wired up starting with `loongarch_parse_expr` which is invoked by the target-specific code:
> 
> https://github.com/bminor/binutils-gdb/blob/binutils-2_38/gas/config/tc-loongarch.c#L593-L659
> 
> Seems not worth it, huh?

Thanks for your explanation. So my understanding is, in a reloc section, several consecutive reloc entries will have the same `r_offset`, and the linker uses these entries plus the stack to calculate the relocation value. In this way we don't have to define new reloc types, we just need to tweak the assembler a little bit so it can generate new combinations of pops and pushes. It that right?

I was trying to figure out whether LoongArch support could be added to the mold linker, and discovered this peculiarity haha.

**Comment by xen0n** [comment]
> Thanks for your explanation. So my understanding is, in a reloc section, several consecutive reloc entries will have the same `r_offset`, and the linker uses these entries plus the stack to calculate the relocation value. In this way we don't have to define new reloc types, we just need to tweak the assembler a little bit so it can generate new combinations of pops and pushes. It that right?

Exactly. If not for obscuring the real intent and the code bloat I think every arch should have converted to this style, obviously the downsides outweigh the benefits.

> I was trying to figure out whether LoongArch support could be added to the mold linker, and discovered this peculiarity haha.

I've been considering doing the same thing the moment mold came out ;-) Better wait until the new scheme is implemented and upstreamed though, this way we won't have to deal with the stack machine at all, so a huge refactoring could be avoided.

**Comment by xen0n** [comment]
BTW, the new scheme as designed by me looks like:

## For immediate loading

- `ABS_LO_UK12 = 64` -- for `ori`
- `ABS_HI12_SJ20 = 65` -- for `lu12i.w`
- `ABS_HI32_SJ20 = 66` -- for `lu32i.d`
- `ABS_HI52_SK12 = 67` -- for `lu52i.d`

## For branches

- `PCREL_SK16 = 68`
    - for other conditional branches
- `PCREL_SD5K16 = 69`
    - `b{eq,ne}z`
    - `bc{eq,ne}z`
- `PCREL_SD10K16 = 70`
    - `b`
    - `bl`

## For PC-relative accesses

- `PCREL_HI12_SJ20= 71`
- `PCREL_LO_UK12 = 72`
- `PCREL_HI32_SJ20 = 73`
- `PCREL_HI52_SK12 = 74`


## For TLS accesses

- `TPREL_LO_UK12 = 75`
- `TPREL_HI12_SJ20 = 76`
- `TPREL_HI32_SJ20 = 77`
- `TPREL_HI52_SK12 = 78`

This is just excerpt from my local notes.

All are of the format `<OPERATION>_<BITFIELD>_<INSN_SLOT>`, where `<INSN_SLOT>` refers to a slot in the instruction word, in the [loongarch-opcodes naming scheme](https://github.com/loongson-community/loongarch-opcodes) so the names are self-explanatory.
